rfcs/005-question-reset.md

```diff
@@ -0,0 +1,87 @@
+
+#### New table: `assessment_question_resets`
+Fields:
+- `id`: unique identifier.
+- `assessment_question_id`: the ID of the aq that this reset applies to.
+- `by`: the uid of the user who performed the reset.
+- `at`: the time that the reset took place.
+- `percentage_score_override`: `NULL` if instructor kept scores of previous attempts, `0` through `100` otherwise.
+
+#### New table: `instance_question_resets`
+
+- `iq`: unique identifier.
+- `assessment_question_reset_id`: the ID of the reset that affected these iqs.
+- `old_iq_id`: the ID of the iq that was replaced in the reset.
+- `new_iq_id`: the ID of the iq that was generated in the reset. 
+
+### User Interface
+
+#### Instructors
+
+Let there be a button labeled `Reset` that appears on the each row in the "questions" tab of an assessment, as well as in each question row under "details" in the "students" tab. When clicked, a dialog appears: 
+- Title: `Reset question`
+- Body: `This will hard reset this question for [all students / uid of student]. All current progress and submissions will be discarded. Designed for cases where a question is being fixed or replaced.`
+- Credit radio select:
+  - Label: `Award credit for previously completed work`
+  - Options:
+    - Default: `Partial (same as already awarded)`
+    - `Custom amount [number input 0-100, default: 100]`
+- Submit options:
+  - `Regenerate`
+  - `Cancel`
+
+If a reset has happened, then in both areas where the Regenerate button can appear, there should be some indicator of this, with information about the time that it happened and which instructor did it. If multiple resets have happened, list them.
+  
+#### Students
+
+On the assessment instance page, next to a question that's been reset, there will be a badge labeled "Reset". Focusing on it gives a popover that reads `This question has been reset by the instructor.`, followed by a link to the reset question and a description of the grading strategy used for their previous attempts:
+> `Credit up to the highest score earned on previous attempts has been awarded. {percentage chosen by instructor}% credit was automatically given, but can still be increased by new attempts.` (exclude last part if percentage == 100)
+
+On a reset instance question's page, there will be an alert containing the same info as the "Reset" popover on the assessment instance page, but with a link to the new question instead of the old one. Assessment navigation buttons ("previous"/"next") don't render in this case, since the instance question no longer exists inside the question order.
+
+### Implementation
+
+Any time that an instructor resets a question, create the relevant entries in `assessment_question_resets` and `instance_question_resets`.
+
+When querying any instance question, do a left exclusive join on `instance_question_resets` to ignore reset questions. 
+
+When a new instance question is generated, automatically set its score based on instructor input.
```

Comment:
This implementation should parallel the "try again" button presented to students with an invalid variant. I'm not sure how to describe entry point here.